### PR TITLE
feat: add keyboard nav and compare mode to Git Graph (#100)

### LINKED ISSUE #100: Git Graph：方向鍵切換 commit（↑/↓、Shift+↑/↓）與 Shift+click 比較兩個 commit
Git Graph 的 commit 清單目前只能用滑鼠單選、沒有鍵盤操作，也無法比較兩個 commit。希望加上：

### 1. ↑ / ↓ — 一般移動
逐列在相鄰 commit 之間上下切換（不分分支，就是清單的上一列 / 下一列），選到的 commit 同步展開右側 detail。

### 2. Shift + ↑ / ↓ — 沿分支路線移動
沿著該 commit 所在分支的線路移動，跳到同一條分支上的上一個 / 下一個 commit（跟著 graph 的分支路徑走，略過其他分支的 commit）。

### 3. Shift + click — 比較兩個 commit
先點選一個 commit，再按住 Shift 點另一列，即顯示這兩個 commit 之間的差異（diff）。

（相關：#94）

### PR DESCRIPTION
**Pull request by mukiwu**
## Summary
- Add Up/Down keyboard navigation between adjacent Git Graph commit rows, clamped at the list boundaries (no wraparound)
- Add Shift+Up/Down navigation that follows a commit's branch line (first-parent chain), auto-paging in more history when the target isn't loaded yet
- Add Shift+click to diff two arbitrary commits against each other in the details panel, with a two-hash header and the AI-explain tab hidden while comparing
- Add two backend commands (`git_commit_range_files`, `git_commit_range_file_diff`) generalizing the existing "commit vs. its first parent" diff to any two commits

Design: `docs/superpowers/specs/2026-07-04-git-graph-keyboard-nav-design.md`
Plan: `docs/superpowers/plans/2026-07-04-git-graph-keyboard-nav.md`

Closes #100

## Test plan
- [x] `pnpm vitest run` — 1156 tests passing
- [x] `pnpm typecheck` — clean
- [x] `cargo test` (src-tauri) — 213 tests passing
- [x] Local build launches without crashing
- [ ] Manual click-through in a real repo with a merge commit (arrow keys, Shift+arrow at a merge, Shift+click compare, exit compare) — recommended before release, not yet done interactively

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/modules/git-graph/GitGraph.tsx [hunk 1]
```diff
@@ -74,6 +77,64 @@ export function GitGraph({
...
+        usePendingGraphSelectionStore.getState().request(parentHash);
+      }
+    } else if (event.key === "ArrowUp" && event.shiftKey) {
+      event.preventDefault();
+      const targetIndex = laneContinuationRowIndex(edges, currentIndex);
+      if (targetIndex !== null) {
+        onSelectCommit(commits[targetIndex], { shiftKey: false });
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mukiwu) [reply src/modules/git-graph/GitGraph.tsx:156]: Applied in 1f52160 — added a `useEffect` that scrolls the active row into view using the layout coordinates already computed by `computeGraphLayout`, since rows here are manually virtualized rather than always-mounted DOM nodes (a plain `scrollIntoView()` target wouldn't exist for off-screen rows). Added two tests covering the scroll-down and already-visible cases.

File: src/modules/git-graph/GitGraphTabContent.tsx [hunk 2]
```diff
@@ -234,6 +234,32 @@ export function GitGraphTabContent() {
...
+          return { mode: "single", commit };
+        }
+        const anchorIndex = commits.findIndex((c) => c.hash === anchor.hash);
+        const commitIndex = commits.findIndex((c) => c.hash === commit.hash);
+        const [from, to] = anchorIndex > commitIndex ? [anchor, commit] : [commit, anchor];
+        return { mode: "compare", from, to };
+      });
+    },
+    [commits],
+  );
```
[the remark this answers is not in the document]
Comment (mukiwu) [reply src/modules/git-graph/GitGraphTabContent.tsx:282]: Applied in 1f52160 — `handleSelectCommit` now uses a functional `setSelection` update that returns the previous state unchanged when the click resolves to the same single commit or the same compare pair, so React bails out of re-rendering and no redundant details/diff fetch fires. Added a regression test asserting `gitCommitDetails` is only called once across two clicks on the same row.

File: src/modules/git-graph/GitGraphTabContent.tsx [hunk 3]
```diff
@@ -234,6 +234,43 @@ export function GitGraphTabContent() {
...
+          prev?.mode === "single" ? prev.commit : prev?.mode === "compare" ? prev.to : null;
+        if (!anchor || anchor.hash === commit.hash) {
+          if (prev?.mode === "single" && prev.commit.hash === commit.hash) {
+            return prev;
+          }
+          return { mode: "single", commit };
+        }
+        const anchorIndex = commits.findIndex((c) => c.hash === anchor.hash);
+        const commitIndex = commits.findIndex((c) => c.hash === commit.hash);
+        const [from, to] = anchorIndex > commitIndex ? [anchor, commit] : [commit, anchor];
```
[the remark this answers is not in the document]
Comment (mukiwu) [reply src/modules/git-graph/GitGraphTabContent.tsx:274]: Applied in 04f01b7 — added an `anchorIndex === -1 || commitIndex === -1` guard that falls back to a plain single selection on the clicked commit, exactly as suggested. Added a regression test that reloads the commit list out from under a stale anchor (simulating a branch switch) and shift-clicks, asserting the result is single-select rather than an arbitrarily-ordered compare pair.

File: src/modules/git-graph/GitGraph.tsx [hunk 4]
```diff
@@ -74,6 +77,86 @@ export function GitGraph({
...
+    const rowTop = layout.y - ROW_HEIGHT / 2;
+    const rowBottom = layout.y + ROW_HEIGHT / 2;
+    const { scrollTop, clientHeight } = container;
+
+    if (rowTop < scrollTop) {
+      container.scrollTop = rowTop;
+    } else if (rowBottom > scrollTop + clientHeight) {
+      container.scrollTop = rowBottom - clientHeight;
+    }
+  }, [activeHash, layouts]);
```
[the remark this answers is not in the document]
Comment (mukiwu) [reply src/modules/git-graph/GitGraph.tsx:193]: Applied in f6d61ee — added a `prevActiveHashRef` and gated the scroll logic on `activeHash !== prevActiveHashRef.current`, exactly as suggested. Added a regression test that pages in more history (new `commits`/`layouts` identity) while the selection stays put and a manual scroll position is set, asserting scrollTop is left untouched.

File: src/modules/git-graph/GitGraph.tsx [hunk 5]
```diff
@@ -74,6 +77,96 @@ export function GitGraph({
...
+    if (activeHash === prevActiveHashRef.current) {
+      return;
+    }
+    prevActiveHashRef.current = activeHash;
+
+    const layout = layouts[activeHash];
+    if (!layout) {
+      return;
+    }
+    const container = scrollRef.current;
```
[the remark this answers is not in the document]
Comment (mukiwu) [reply src/modules/git-graph/GitGraph.tsx:183]: Applied in aab33b4 — moved `prevActiveHashRef.current = activeHash` to after the `layouts[activeHash]` existence check, exactly as suggested, so a hash change with no layout yet doesn't get silently marked as handled. Added a regression test that changes selection to a not-yet-loaded commit, confirms no scroll happens, then loads it in and confirms the scroll finally fires instead of being skipped.

File: src/modules/git-graph/CommitDetailsPanel.tsx [hunk 6]
```diff
@@ -376,25 +410,27 @@ export function CommitDetailsPanel({ repo, commit, onClose, labels }: CommitDeta
...
               <div className="min-h-0 flex-1">
-                {tab === "diff" ? (
+                {activeTab === "diff" ? (
                   <DiffView lines={diffLines} emptyLabel={labels.noDiff} />
                 ) : (
                   <DiffExplain
-                    key={`${commit.hash}|${selectedFile}`}
-                    commitHash={commit.hash}
+                    key={`${singleCommit!.hash}|${selectedFile}`}
+                    commitHash={singleCommit!.hash}
```
[the remark this answers is not in the document]
Comment (mukiwu) [reply src/modules/git-graph/CommitDetailsPanel.tsx]: Applied in 7c2de8a — switched both `singleCommit!.hash` uses to `singleCommit?.hash ?? ""` optional chaining, exactly as suggested. Agreed this is safer defense-in-depth even though `activeTab` currently guarantees this branch only renders in single mode.

File: src/modules/git-graph/GitGraph.tsx [hunk 7]
```diff
@@ -74,6 +77,101 @@ export function GitGraph({
...
+      if (targetIndex !== currentIndex) {
+        onSelectCommit(commits[targetIndex], { shiftKey: false });
+      }
+    } else if (event.key === "ArrowUp" && !event.shiftKey) {
+      event.preventDefault();
+      const targetIndex = Math.max(currentIndex - 1, 0);
+      if (targetIndex !== currentIndex) {
+        onSelectCommit(commits[targetIndex], { shiftKey: false });
+      }
+    } else if (event.key === "ArrowDown" && event.shiftKey) {
```
[the remark this answers is not in the document]
Comment (mukiwu) [reply src/modules/git-graph/GitGraph.tsx:132]: Applied in b90ba84, extended to full spec fidelity — added an `isComparing` check to all four branches (not just the two plain-arrow ones in the suggestion), so Shift+ArrowDown at a root commit and Shift+ArrowUp at the newest commit on a lane also collapse out of compare mode instead of silently no-oping. Added four regression tests covering each boundary case in compare mode.
